indra/llcharacter/llik.cpp

```diff
@@ -483,38 +478,117 @@ LLQuaternion LLIK::TwistLimitedCone::computeAdjustedLocalRot(const LLQuaternion&
     return adjusted_local_rot;
 }
 
-LLQuaternion LLIK::TwistLimitedCone::minimizeTwist(const LLQuaternion& joint_local_rot) const
+#ifdef DEBUG_LLIK_UNIT_TESTS
+void LLIK::TwistLimitedCone::dumpConfig() const
 {
-    // Compute the swing and combine with default twist
-    // which is the midpoint of the twist range.
-    LLQuaternion mid_twist;
-    mid_twist.setAngleAxis(0.5f * (mMinTwist + mMaxTwist), mForward);
+    LL_INFOS("debug") << "{'type':'TwistLimitedCone'"
+        << ",'forward':(" << mForward.mV[VX] << "," << mForward.mV[VY] << "," << mForward.mV[VZ] << ")"
+        << ",'cone_angle':" << std::atan2(mSinConeAngle, mCosConeAngle)
+        << ",'min_twist':" << mMinTwist
+        << ",'max_twist':" << mMaxTwist << "}" << LL_ENDL;
+}
+#endif
 
-    // joint_local_rot = mid_twist * swing
-    LLQuaternion new_local_rot = mid_twist;
+LLIK::ShoulderConstraint::ShoulderConstraint()
+{
+    mType = LLIK::Constraint::SHOULDER_CONSTRAINT;
+}
 
-    LLVector3 joint_forward = mForward * joint_local_rot;
-    LLVector3 swing_axis = mForward % joint_forward;
-    constexpr F32 MIN_SWING_AXIS_LENGTH = 1.0e-3f;
-    if (swing_axis.length() > MIN_SWING_AXIS_LENGTH)
+LLIK::ShoulderConstraint::ShoulderConstraint(LLSD &parameters):
+    LLIK::Constraint(LLIK::Constraint::SHOULDER_CONSTRAINT, parameters)
+{
+}
```

Comment:
Shoulder constraint doesn't have any private data members yet.  This may change later but for the now all it does is try to keep the elbow bend pointed down.  This is done in the `enforce()` override below, and the implementation of its `computeAdjustedLocalRot()` method is a no-op.